ledger/catchpointwriter.go

```diff
@@ -0,0 +1,299 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"database/sql"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/algorand/go-codec/codec"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/util/db"
+)
+
+// balancesChunkReadSize defines the number of accounts that would be stored in each chunk in the catchpoint file.
+// note that the last chunk would typically be less than this number.
+const balancesChunkReadSize = 512
+
+// catchpointFileVersion is the catchpoint file version
+const catchpointFileVersion = uint64(0200)
+
+// catchpointWriter is the struct managing the persistance of accounts data into the catchpoint file.
+// it's designed to work in a step fashion : a caller will call the WriteStep method in a loop until
+// the writing is complete. It migth take multiple steps until the operation is over, and the caller
+// has the option of throtteling the CPU utilization in between the calls.
+type catchpointWriter struct {
+	hasher            hash.Hash
+	innerWriter       io.WriteCloser
+	dbr               db.Accessor
+	filePath          string
+	file              *os.File
+	gzip              *gzip.Writer
+	tar               *tar.Writer
+	headerWritten     bool
+	balancesOffset    int
+	balancesChunk     catchpointFileBalancesChunk
+	fileHeader        *catchpointFileHeader
+	balancesChunkNum  uint64
+	writtenBytes      int64
+	blocksRound       basics.Round
+	blockHeaderDigest crypto.Digest
+	label             string
+}
+
+type encodedBalanceRecord struct {
+	_struct     struct{}  `codec:",omitempty,omitemptyarray"`
+	Address     []byte    `codec:"pk"`
+	AccountData codec.Raw `codec:"ad"`
+}
+
+type catchpointFileHeader struct {
+	_struct           struct{}      `codec:",omitempty,omitemptyarray"`
+	Version           uint64        `codec:"version"`
+	BalancesRound     basics.Round  `codec:"balancesRound"`
```

Comment:
Strictly speaking - No. We might be able to avoid storing both BalancesRound as well as BlocksRound. However, it helps to have all the data in one place. Given that the storage cost is not really meaningful, I have included few additional fields here which I think could become useful. For instance Catchpoint or BlockHeaderDigest.
These could be used by external tools to verify the correctness and metadata of this file.
Note that the file is designed to be binary *identical* across machines, which is why it doesn't contain any time signature or machine-specific metadata.